Restore the difference between "live" and "up"

konklone opened the pull request:
In site-inspector 1.X, "live" meant a 2XX response code. It would mean `!!response`, where `response` would be truthy if `response.success?` was true, and `false` if the response was nil or failed.

In 1.0.2, I added "up", which meant: "Is this domain/endpoint reachable over HTTP at all?" Even an endpoint that returned a 500 response code was "up" in the sense that the site responded to connections.

There are various cases where it's valuable to measure whether a domain responds to HTTP at all, and to distinguish an unreachable endpoint from an endpoint which just has a [404 page lying around](https://www.nist.gov/).

In 2.0, "live" got renamed to "up", which changed the intended meaning of "up". This pull request separates their definitions again:

```ruby
    # Does the endpoint return a 2xx or 3xx response code?
    def live?
      response && response_code.start_with?("2") || response_code.start_with?("3")
    end

    def dead?
      !live?
    end

    # Does the domain respond to HTTP?
    def up?
      response?
    end

    def down?
      !up?
    end
```

```ruby
    def response?
       response.code != 0 && !timed_out?
    end
```

The domain-wide calculation is still the same: a domain is "live" if any of its endpoints are "live", and "up" if any of its endpoints are "up". The tests are updated as well.

In this pull request, I left alone the downstream decisions that depend on live/up, like whether an endpoint is "canonical". Those will likely be changed in a subsequent PR, because the canonicality definitions I created for 1.0.2 were sometimes premised on an endpoint being reachable, sometimes premised on an endpoint having a 20X.

benbalter commented:
I like the distinction, but the terminology is confusing. Would `response?` accurate describe `up?`.

konklone commented:
@benbalter Yes, the terminology is confusing. `response?` is closer. How about `responds?`?

konklone commented:
Okay @benbalter, I moved `live?` back to `up?`, and now have `responds?` for whether an endpoint responds.

I also removed `down?`, and I found it to be more confusing to parse logic statements that used it instead of `!up?`, and it wasn't used during JSON output.

I created `responds?` by simply renaming `reponse?` and updating the tests.

benbalter commented on lib/site-inspector/endpoint.rb:
```diff
@@ -74,8 +69,9 @@ def up?
+    # Does the server respond at all?
+    def responds?
+       response.code != 0 && !timed_out?
     end
```
Would it be worth adding an alias here so that we have one less breaking change?

benbalter commented on spec/site_inspector_domain_spec.rb:
```diff
@@ -68,7 +68,29 @@
+    end
+  end
+
+  context "up" do
```
Looks like we have two "up" contexts now. Was that intentional?

konklone replied:
Nope - I wonder if it obscured the test run, too. Who will test the tests???

benbalter commented on spec/site_inspector_domain_spec.rb:
```diff
@@ -68,7 +68,29 @@
+  end
+
+  context "up" do
+    it "considers a domain up if at least one endpoint is up" do
```
At least on the name, this looks like a duplicate test?

konklone replied:
Yeah, I'll clean this up. Sorry for the premature merge.

benbalter commented:
@konklone may also be worth updating the README to explain the difference between `up?` and `response?`.
